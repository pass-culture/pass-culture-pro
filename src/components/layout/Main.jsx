import classnames from 'classnames'
import get from 'lodash.get'
import {
  Icon,
  Modal,
  resetForm,
  showNotification,
  Spinner,
  withBlock,
} from 'pass-culture-shared'
import React, { Component, Fragment } from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import { NavLink } from 'react-router-dom'
import ReactTooltip from 'react-tooltip'
import { compose } from 'redux'

import Header from './Header'
import Notification from './Notification'

class Main extends Component {
  constructor() {
    super()
    this.state = {
      loading: false,
    }
  }

  static defaultProps = {
    Tag: 'main',
  }

  handleDataFail = (state, action) => {
    const { dispatch, payload } = action
    this.setState({ loading: false })

    dispatch(
      showNotification({
        type: 'danger',
        text: get(payload, 'errors.0.global') || 'Erreur de chargement',
      })
    )
  }

  handleDataRequest = () => {
    if (this.props.handleDataRequest) {
      // possibility of the handleDataRequest to return
      // false in order to not trigger the loading
      this.setState({
        loading: true,
      })
      this.props.handleDataRequest(this.handleDataSuccess, this.handleDataFail)
    }
  }

  handleDataSuccess = () => {
    this.setState({
      loading: false,
    })
  }

  componentDidMount() {
    this.props.user && this.handleDataRequest()
  }

  componentDidUpdate(prevProps) {
    const userChanged = !prevProps.user && this.props.user // User just loaded
    const searchChanged =
      this.props.location.search !== prevProps.location.search

    if (userChanged || searchChanged) {
      this.handleDataRequest()
    }
  }

  componentWillUnmount() {
    this.unblock && this.unblock()
    this.props.dispatch(resetForm())
  }

  render() {
    const {
      backTo,
      children,
      fullscreen,
      header,
      name,
      redBg,
      Tag,
      whiteHeader,
      withLoading,
    } = this.props
    const { loading } = this.state
    const footer = [].concat(children).find(e => e && e.type === 'footer')
    const $content = []
      .concat(children)
      .filter(e => e && e.type !== 'header' && e.type !== 'footer')

    return (
      <Fragment>
        {!fullscreen && <Header whiteHeader={whiteHeader} {...header} />}
        <ReactTooltip
          className="flex-center items-center"
          delayHide={500}
          effect="solid"
          html
        />
        <Tag
          className={classnames({
            page: true,
            [`${name}-page`]: true,
            'with-header': Boolean(header),
            'red-bg': redBg,
            'white-header': whiteHeader,
            container: !fullscreen,
            fullscreen,
            loading,
          })}>
          {fullscreen ? (
            <Fragment>
              <Notification isFullscreen />
              {$content}
            </Fragment>
          ) : (
            <div className="columns is-gapless">
              <div className="page-content column is-10 is-offset-1">
                <Notification />
                <div
                  className={classnames('after-notification-content', {
                    'with-padding': backTo,
                  })}>
                  {backTo && (
                    <NavLink
                      to={backTo.path}
                      className="back-button has-text-primary has-text-weight-semibold">
                      <Icon svg="ico-back" />
                      {` ${backTo.label}`}
                    </NavLink>
                  )}
                  <div className="main-content">{$content}</div>
                  {withLoading && loading && <Spinner />}
                </div>
              </div>
            </div>
          )}
          {footer}
        </Tag>
        <Modal key="modal" />
      </Fragment>
    )
  }
}

function mapStateToProps(state) {
  return {
    user: state.user,
  }
}

export default compose(
  withRouter,
  withBlock,
  connect(mapStateToProps)
)(Main)
