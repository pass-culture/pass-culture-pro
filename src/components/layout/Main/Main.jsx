import classnames from 'classnames'
import get from 'lodash.get'
import { Icon, Modal, resetForm, Spinner } from 'pass-culture-shared'
import PropTypes from 'prop-types'
import React, { PureComponent, Fragment } from 'react'
import { NavLink } from 'react-router-dom'
import ReactTooltip from 'react-tooltip'

import ActionsBar from '../ActionsBar/'
import HeaderContainer from 'components/layout/Header/HeaderContainer'
import NotificationV1Container from 'components/layout/NotificationV1/NotificationV1Container'
import { showNotificationV1 } from 'store/reducers/notificationReducer'

class Main extends PureComponent {
  constructor() {
    super()
    this.state = {
      loading: false,
    }
  }

  componentDidMount() {
    const { currentUser } = this.props
    if (currentUser) {
      this.handleDataRequest()
    }
  }

  componentDidUpdate(prevProps) {
    const { currentUser, location } = this.props
    const { search } = location
    const userChanged = !prevProps.currentUser && currentUser // User just loaded
    const searchChanged = search !== prevProps.location.search

    if (userChanged || searchChanged) {
      this.handleDataRequest()
    }
  }

  componentWillUnmount() {
    this.unblock && this.unblock()
    const { dispatch } = this.props
    dispatch(resetForm())
  }

  handleDataSuccess = () => {
    this.setState({
      loading: false,
    })
  }

  handleDataRequest = () => {
    const { handleDataRequest } = this.props
    if (handleDataRequest) {
      // possibility of the handleDataRequest to return
      // false in order to not trigger the loading
      this.setState({
        loading: true,
      })
      handleDataRequest(this.handleDataSuccess, this.handleDataFail)
    }
  }

  handleDataFail = (state, action) => {
    const { dispatch, payload } = action
    this.setState({ loading: false })

    dispatch(
      showNotificationV1({
        type: 'danger',
        text: get(payload, 'errors.0.global') || 'Erreur de chargement',
      })
    )
  }

  render() {
    const {
      PageActionsBar,
      backTo,
      children,
      fullscreen,
      header,
      name,
      redBg,
      Tag,
      whiteHeader,
      withLoading,
    } = this.props
    const { loading } = this.state
    const footer = [].concat(children).find(e => e && e.type === 'footer')
    const $content = []
      .concat(children)
      .filter(e => e && e.type !== 'header' && e.type !== 'footer')

    return (
      <Fragment>
        {!fullscreen && (
          <HeaderContainer
            whiteHeader={whiteHeader}
            {...header}
          />
        )}
        <ReactTooltip
          className="flex-center items-center"
          delayHide={500}
          effect="solid"
          html
        />
        <Tag
          className={classnames({
            page: true,
            [`${name}-page`]: true,
            'with-header': Boolean(header),
            'red-bg': redBg,
            'white-header': whiteHeader,
            container: !fullscreen,
            fullscreen,
            loading,
          })}
        >
          {fullscreen ? (
            <Fragment>
              <NotificationV1Container isFullscreen />
              {$content}
            </Fragment>
          ) : (
            <div className="columns is-gapless">
              <div className="page-content column is-10 is-offset-1">
                <NotificationV1Container />
                <div
                  className={classnames('after-notification-content', {
                    'with-padding': backTo,
                  })}
                >
                  {backTo && (
                    <NavLink
                      className="back-button has-text-primary has-text-weight-semibold"
                      to={backTo.path}
                    >
                      <Icon svg="ico-back" />
                      {` ${backTo.label}`}
                    </NavLink>
                  )}
                  <div className="main-content">
                    {$content}
                  </div>
                  {withLoading && loading && <Spinner />}
                </div>
              </div>
            </div>
          )}
          {footer}
        </Tag>
        <Modal key="modal" />
        { PageActionsBar && <ActionsBar><PageActionsBar/></ActionsBar>}

      </Fragment>
    )
  }
}

Main.defaultProps = {
  Tag: 'main',
  PageActionsBar: null,
  backTo: null,
  fullscreen: false,
  handleDataRequest: null,
  header: {},
  redBg: null,
  whiteHeader: null,
  withLoading: null,
}

Main.propTypes = {
  Tag: PropTypes.string,
  PageActionsBar: PropTypes.elementType,
  backTo: PropTypes.shape(),
  children: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  currentUser: PropTypes.shape().isRequired,
  dispatch: PropTypes.func.isRequired,
  fullscreen: PropTypes.bool,
  handleDataRequest: PropTypes.func,
  header: PropTypes.shape(),
  location: PropTypes.shape().isRequired,
  name: PropTypes.string.isRequired,
  redBg: PropTypes.string,
  whiteHeader: PropTypes.string,
  withLoading: PropTypes.bool,
}

export default Main
