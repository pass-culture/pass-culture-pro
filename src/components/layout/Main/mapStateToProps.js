import { selectCurrentUser } from 'with-react-redux-login'

function mapStateToProps(state) {
  return {
    currentUser: selectCurrentUser(state),
  }
}

export default mapStateToProps
