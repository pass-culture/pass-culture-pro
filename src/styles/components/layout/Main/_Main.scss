.main-content {
  min-height: rem(800px)
}
