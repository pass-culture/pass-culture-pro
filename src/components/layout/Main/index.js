import { withBlock } from 'pass-culture-shared'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import { compose } from 'redux'

import Main from './Main'
import mapStateToProps from './mapStateToProps'

export default compose(
  withRouter,
  withBlock,
  connect(mapStateToProps)
)(Main)
