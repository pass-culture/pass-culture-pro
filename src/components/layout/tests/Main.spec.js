import React from 'react'
import { shallow } from 'enzyme'

import { RawMain } from '../Main'

describe('src | components | layout | Main', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      // given
      const props = {
        dispatch: jest.fn(),
        history: {},
        location: {},
        name: 'foo',
      }

      // when
      const wrapper = shallow(
        <RawMain {...props}>
          <div>foo</div>
        </RawMain>
      )

      // then
      expect(wrapper).toBeDefined()
      expect(wrapper).toMatchSnapshot()
    })
  })
})
