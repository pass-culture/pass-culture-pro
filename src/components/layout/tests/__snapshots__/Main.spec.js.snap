// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | layout | Main snapshot should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RawMain
    Tag="main"
    dispatch={[MockFunction]}
    history={Object {}}
    location={Object {}}
    name="foo"
  >
    <div>
      foo
    </div>
  </RawMain>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <withRouter(Connect(RawHeader)) />,
        <ReactTooltip
          className="flex-center items-center"
          clickable={false}
          delayHide={500}
          effect="solid"
          html={true}
          insecure={true}
          resizeHide={true}
          wrapper="div"
        />,
        <main
          className="page foo-page container"
        >
          <div
            className="columns is-gapless"
          >
            <div
              className="page-content column is-10 is-offset-1"
            >
              <Connect(Notification) />
              <div
                className="after-notification-content"
              >
                <div
                  className="main-content"
                >
                  <div>
                    foo
                  </div>
                </div>
              </div>
            </div>
          </div>
        </main>,
        <withRouter(Connect(Modal)) />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "whiteHeader": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "className": "flex-center items-center",
          "clickable": false,
          "delayHide": 500,
          "effect": "solid",
          "html": true,
          "insecure": true,
          "resizeHide": true,
          "wrapper": "div",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="columns is-gapless"
            >
              <div
                className="page-content column is-10 is-offset-1"
              >
                <Connect(Notification) />
                <div
                  className="after-notification-content"
                >
                  <div
                    className="main-content"
                  >
                    <div>
                      foo
                    </div>
                  </div>
                </div>
              </div>
            </div>,
            undefined,
          ],
          "className": "page foo-page container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="page-content column is-10 is-offset-1"
              >
                <Connect(Notification) />
                <div
                  className="after-notification-content"
                >
                  <div
                    className="main-content"
                  >
                    <div>
                      foo
                    </div>
                  </div>
                </div>
              </div>,
              "className": "columns is-gapless",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Connect(Notification) />,
                  <div
                    className="after-notification-content"
                  >
                    <div
                      className="main-content"
                    >
                      <div>
                        foo
                      </div>
                    </div>
                  </div>,
                ],
                "className": "page-content column is-10 is-offset-1",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      undefined,
                      <div
                        className="main-content"
                      >
                        <div>
                          foo
                        </div>
                      </div>,
                      undefined,
                    ],
                    "className": "after-notification-content",
                  },
                  "ref": null,
                  "rendered": Array [
                    undefined,
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": Array [
                          <div>
                            foo
                          </div>,
                        ],
                        "className": "main-content",
                      },
                      "ref": null,
                      "rendered": Array [
                        Object {
                          "instance": null,
                          "key": undefined,
                          "nodeType": "host",
                          "props": Object {
                            "children": "foo",
                          },
                          "ref": null,
                          "rendered": "foo",
                          "type": "div",
                        },
                      ],
                      "type": "div",
                    },
                    undefined,
                  ],
                  "type": "div",
                },
              ],
              "type": "div",
            },
            "type": "div",
          },
          undefined,
        ],
        "type": "main",
      },
      Object {
        "instance": null,
        "key": "modal",
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <withRouter(Connect(RawHeader)) />,
          <ReactTooltip
            className="flex-center items-center"
            clickable={false}
            delayHide={500}
            effect="solid"
            html={true}
            insecure={true}
            resizeHide={true}
            wrapper="div"
          />,
          <main
            className="page foo-page container"
          >
            <div
              className="columns is-gapless"
            >
              <div
                className="page-content column is-10 is-offset-1"
              >
                <Connect(Notification) />
                <div
                  className="after-notification-content"
                >
                  <div
                    className="main-content"
                  >
                    <div>
                      foo
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </main>,
          <withRouter(Connect(Modal)) />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "whiteHeader": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "className": "flex-center items-center",
            "clickable": false,
            "delayHide": 500,
            "effect": "solid",
            "html": true,
            "insecure": true,
            "resizeHide": true,
            "wrapper": "div",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="columns is-gapless"
              >
                <div
                  className="page-content column is-10 is-offset-1"
                >
                  <Connect(Notification) />
                  <div
                    className="after-notification-content"
                  >
                    <div
                      className="main-content"
                    >
                      <div>
                        foo
                      </div>
                    </div>
                  </div>
                </div>
              </div>,
              undefined,
            ],
            "className": "page foo-page container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  className="page-content column is-10 is-offset-1"
                >
                  <Connect(Notification) />
                  <div
                    className="after-notification-content"
                  >
                    <div
                      className="main-content"
                    >
                      <div>
                        foo
                      </div>
                    </div>
                  </div>
                </div>,
                "className": "columns is-gapless",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <Connect(Notification) />,
                    <div
                      className="after-notification-content"
                    >
                      <div
                        className="main-content"
                      >
                        <div>
                          foo
                        </div>
                      </div>
                    </div>,
                  ],
                  "className": "page-content column is-10 is-offset-1",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {},
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": Array [
                        undefined,
                        <div
                          className="main-content"
                        >
                          <div>
                            foo
                          </div>
                        </div>,
                        undefined,
                      ],
                      "className": "after-notification-content",
                    },
                    "ref": null,
                    "rendered": Array [
                      undefined,
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": Array [
                            <div>
                              foo
                            </div>,
                          ],
                          "className": "main-content",
                        },
                        "ref": null,
                        "rendered": Array [
                          Object {
                            "instance": null,
                            "key": undefined,
                            "nodeType": "host",
                            "props": Object {
                              "children": "foo",
                            },
                            "ref": null,
                            "rendered": "foo",
                            "type": "div",
                          },
                        ],
                        "type": "div",
                      },
                      undefined,
                    ],
                    "type": "div",
                  },
                ],
                "type": "div",
              },
              "type": "div",
            },
            undefined,
          ],
          "type": "main",
        },
        Object {
          "instance": null,
          "key": "modal",
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
